From d930e86fce26bb4db9d9f00a9d1890ffc8cf0c43 Mon Sep 17 00:00:00 2001
From: Hunt Xu <mhuntxu@gmail.com>
Date: Tue, 5 Sep 2017 16:19:48 +0800
Subject: [PATCH 108/112] Sync #118 from devel to testing

PPTP: allow the same username used by different tenants

Fixes: redmine #10809

Signed-off-by: Hunt Xu <mhuntxu@gmail.com>
(cherry picked from commit 77b18f890e34fac82d64d9ec868ac7344c67b2bd)
Signed-off-by: Hunt Xu <mhuntxu@gmail.com>
---
 neutron/db/vpn/vpn_db.py | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/neutron/db/vpn/vpn_db.py b/neutron/db/vpn/vpn_db.py
index d50d4bd6f..314f748b7 100644
--- a/neutron/db/vpn/vpn_db.py
+++ b/neutron/db/vpn/vpn_db.py
@@ -659,9 +659,10 @@ class VPNPluginDb(vpnaas.VPNPluginBase, base_db.CommonDbMixin):
                                pptp_credential['associations']]}
         return self._fields(res, fields)
 
-    def _username_already_exists(self, context, username):
-        query = self._model_query(context, PPTPCredential)
-        return len(query.filter_by(username=username).all()) > 0
+    def _username_already_exists(self, context, tenant_id, username):
+        credentials = self._model_query(context, PPTPCredential).filter_by(
+            tenant_id=tenant_id, username=username).all()
+        return len(credentials) > 0
 
     def _create_port_for_vpnservice(self, context,
                                     vpnservice_id, pptp_credential_id):
@@ -695,7 +696,7 @@ class VPNPluginDb(vpnaas.VPNPluginBase, base_db.CommonDbMixin):
         pptp_credential_id = uuidutils.generate_uuid()
         with context.session.begin(subtransactions=True):
             username = pptp_credential['username']
-            if self._username_already_exists(context, username):
+            if self._username_already_exists(context, tenant_id, username):
                 raise vpnaas.PPTPUsernameAlreadyExists(username=username)
             pptp_credential_db = PPTPCredential(
                 id=pptp_credential_id,
-- 
2.11.0 (Apple Git-81)

